File: app/src/main/java/com/absinthe/kage/device/DeviceManager.java
package com.absinthe.kage.device;

import android.bluetooth.BluetoothAdapter;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;

import com.absinthe.kage.connect.Const;
import com.absinthe.kage.connect.IProxy;
import com.absinthe.kage.connect.protocol.Config;
import com.absinthe.kage.connect.tcp.KageSocket;
import com.absinthe.kage.device.cmd.PromptPhoneConnectedCommand;
import com.absinthe.kage.device.model.DeviceConfig;
import com.absinthe.kage.device.model.DeviceInfo;
import com.absinthe.kage.utils.NetUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class DeviceManager extends KageObservable implements LifecycleObserver {
    private static final String TAG = DeviceManager.class.getSimpleName();
    private static final int DEFAULT_CONNECT_TIMEOUT = 10 * 1000;
    private final byte[] LOCK = new byte[0];

    public enum Singleton {
        INSTANCE;
        private DeviceManager instance;

        Singleton() {
            instance = new DeviceManager();
        }

        public DeviceManager getInstance() {
            return instance;
        }
    }

    public enum ConnectFailedReason {
        CONNECT_ERROR_CODE_CONNECT_UNKNOWN, //未知原因
        CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE,  //远端 IP 或 Port 不可到达
        CONNECT_ERROR_CODE_CONNECT_TIMEOUT, //连接超时
        CONNECT_ERROR_CODE_HAND_SHAKE_UNDONE    //协议握手未完成
    }

    public enum Result {
        RESULT_START_MONITOR_DEVICE_SUCCESS,
        RESULT_START_MONITOR_DEVICE_FAILED_NOT_INIT,
        RESULT_START_MONITOR_DEVICE_FAILED_UNAUTHORIZED,
        RESULT_START_MONITOR_DEVICE_FAILED_INTERNAL_ERROR
    }

    private DeviceConfig mConfig;
    private DeviceScanner mDeviceScanner;
    private List<IDeviceObserver> observers = new ArrayList<>();
    private List<IProxy> mProxyList = new ArrayList<>();
    private Handler mHandler = new Handler(Looper.getMainLooper());
    private ConnectState mConnectState = null;
    private String mCurrentDeviceKey = null;
    private String delayToConnectDeviceInfoIp = null;

    private DeviceManager() {
        setConnectState(new StateIdle());
    }

    public void init() {
        if (mDeviceScanner == null) {
            mDeviceScanner = new DeviceScanner();
        }

        DeviceConfig config = new DeviceConfig();
        BluetoothAdapter defaultAdapter = BluetoothAdapter.getDefaultAdapter();

        if (defaultAdapter != null) {
            config.name = defaultAdapter.getName();
        } else {
            config.name = android.os.Build.MODEL;
        }

        config.uuid = UUID.randomUUID().toString();
        config.localHost = NetUtils.getLocalAddress();
        config.broadcastHostInWifi = Const.BROADCAST_IP_IN_WIFI;
        config.broadcastHostInAp = Const.BROADCAST_IP_IN_AP;
        config.broadcastMonitorPort = Config.PORT;
        config.broadcastPort = Config.PORT;
        mConfig = config;
        mDeviceScanner.setConfig(mConfig);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void release() {
        if (isConnected()) {
            disConnectDevice();
        }
        if (mDeviceScanner != null) {
            mDeviceScanner.stopScan();
            mDeviceScanner = null;
        }
        mHandler.removeCallbacksAndMessages(null);
    }

    public void startMonitorDevice() {
        startMonitorDevice(6000);
    }

    public int getScanPeriod() {
        if (!checkConfiguration()) {
            return -1;
        }
        return mDeviceScanner.getScanPeriod();
    }

    public DeviceConfig getConfig() {
        return mConfig;
    }

    /**
     * 开始监测设备
     *
     * @param period period
     * @return resultCode
     */
    public int startMonitorDevice(int period) {
        if (!checkConfiguration()) {
            return Result.RESULT_START_MONITOR_DEVICE_FAILED_NOT_INIT.ordinal();
        }
        boolean b = mDeviceScanner.startScan(period, new DeviceScanner.IScanCallback() {
            @Override
            public void onDeviceOnline(final Device device) {
                mHandler.post(() -> notifyFindDevice(device));
            }

            @Override
            public void onDeviceOffline(final Device device) {
                mHandler.post(() -> notifyLostDevice(device));
            }

            @Override
            public void onDeviceInfoChanged(final Device device) {
                mHandler.post(() -> notifyDeviceInfoChanged(device));
            }

            @Override
            public void onDeviceNotice(final Device device) {
                mHandler.post(() -> notifyDeviceNotice(device));
            }
        });
        if (b) {
            return Result.RESULT_START_MONITOR_DEVICE_SUCCESS.ordinal();
        } else {
            return Result.RESULT_START_MONITOR_DEVICE_FAILED_INTERNAL_ERROR.ordinal();
        }
    }

    /**
     * 停止监测设备
     */
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void stopMonitorDevice() {
        if (!checkConfiguration()) {
            return;
        }
        mDeviceScanner.stopScan();
    }

    public List<DeviceInfo> getDeviceInfoList() {
        if (!checkConfiguration()) {
            Log.d(TAG, "checkConfiguration() return false");
            return new ArrayList<>();
        }
        List<DeviceInfo> deviceInfoList = new ArrayList<>();
        Map<String, Device> devices = mDeviceScanner.getDevices();
        for (Device device : devices.values()) {
            DeviceInfo deviceInfo = device.getDeviceInfo();
            if (device.getState() != DeviceInfo.STATE_IDLE) {
                deviceInfoList.add(0, deviceInfo);
            } else {
                deviceInfoList.add(deviceInfo);
            }
        }
        return deviceInfoList;
    }

    @Override
    public synchronized void register(IDeviceObserver observer) {
        if (observer == null) {
            return;
        }
        boolean hasRegistered = false;
        int size = observers.size();
        for (int i = 0; i < size; i++) {
            if (observer.equals(observers.get(i))) {
                hasRegistered = true;
                break;
            }
        }
        if (!hasRegistered) {
            observers.add(observer);
        }
    }

    @Override
    public synchronized void unregister(IDeviceObserver observer) {
        if (observer == null) {
            return;
        }
        observers.remove(observer);
    }

    @Override
    protected void notifyFindDevice(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onFindDevice(DeviceInfo);
        }
    }

    @Override
    protected void notifyLostDevice(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onLostDevice(DeviceInfo);
        }
    }

    @Override
    protected void notifyDeviceConnected(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo deviceInfo = device.getDeviceInfo();
            deviceInfo.setState(DeviceInfo.STATE_CONNECTED);
            observer.onDeviceConnected(deviceInfo);
        }
    }

    @Override
    protected void notifyDeviceDisconnect(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo deviceInfo = device.getDeviceInfo();
            deviceInfo.setState(DeviceInfo.STATE_IDLE);
            observer.onDeviceDisConnect(deviceInfo);
        }
    }

    @Override
    protected void notifyDeviceConnectFailed(Device device, int errorCode, String errorMessage) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onDeviceConnectFailed(DeviceInfo, errorCode, errorMessage);
        }
    }

    private void notifyDeviceInfoChanged(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onDeviceInfoChanged(DeviceInfo);
        }
    }

    private void notifyDeviceConnecting(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onDeviceConnecting(DeviceInfo);
        }
    }

    private void notifyDeviceNotice(Device device) {
        IDeviceObserver[] localObservers;
        synchronized (this) {
            localObservers = observers.toArray(new IDeviceObserver[0]);
        }
        for (IDeviceObserver observer : localObservers) {
            DeviceInfo DeviceInfo = device.getDeviceInfo();
            observer.onDeviceNotice(DeviceInfo);
        }
    }

    /**
     * 连接指定的设备，已连接事件会通知Observer。
     */
    public void connectDevice(DeviceInfo deviceInfo) {
        synchronized (LOCK) {
            if (checkConfiguration()) {
                mConnectState.connect(deviceInfo);
            }
        }
    }

    /**
     * @param deviceInfo Device info
     * @param timeout    连接超时时间，单位为毫秒
     */
    public synchronized void connectDevice(DeviceInfo deviceInfo, int timeout) {
        if (checkConfiguration()) {
            mConnectState.connect(deviceInfo, timeout);
        }
    }

    /**
     * 如果已连接设备，断开当前连接。已断开事件会通知Observer
     */
    public void disConnectDevice() {
        mConnectState.disConnect();
    }

    private void setConnectState(ConnectState connectState) {
        mConnectState = connectState;
    }

    public DeviceInfo getCurrentDeviceInfo() {
        Device device = getCurrentDevice();
        if (device == null) return null;
        return device.getDeviceInfo();
    }

    @Nullable
    private Device getCurrentDevice() {
        if (!checkConfiguration()) {
            return null;
        }
        Device device;
        synchronized (LOCK) {
            if (null == mCurrentDeviceKey) {
                return null;
            }
            device = mDeviceScanner.queryDevice(mCurrentDeviceKey);
        }
        return device;
    }

    public boolean isConnected() {
        Device currentDevice = getCurrentDevice();
        if (null != currentDevice) {
            return currentDevice.isConnected();
        }
        return false;
    }

    public DeviceInfo onlineDevice(DeviceInfo deviceInfo) {
        if (!checkConfiguration()) {
            return null;
        }
        return mDeviceScanner.onlineDevice(deviceInfo);
    }

    public boolean offlineDevice(DeviceInfo deviceInfo) {
        if (!checkConfiguration()) {
            return false;
        }
        return mDeviceScanner.offlineDevice(deviceInfo);
    }

    private boolean checkConfiguration() {
        return this.mConfig != null && mDeviceScanner != null;
    }

    private class StateIdle implements ConnectState {

        @Override
        public void connect(DeviceInfo deviceInfo) {
            connect(deviceInfo, DEFAULT_CONNECT_TIMEOUT);
        }

        @Override
        public void connect(DeviceInfo deviceInfo, int timeout) {
            Log.d(TAG, "StateIdle connect");
            if (deviceInfo == null) {
                return;
            }
            if (timeout < 0) {
                timeout = DEFAULT_CONNECT_TIMEOUT;
            }
            String key = deviceInfo.getIp();
            final Device device = mDeviceScanner.queryDevice(key);
            if (device == null) {
                return;
            }
            synchronized (LOCK) {
                mCurrentDeviceKey = key;
            }
            setConnectState(new StateConnecting(device));
            device.setConnectCallback(new Device.IConnectCallback() {
                @Override
                public void onConnectedFailed(final int code, Exception e) {
                    Log.d(TAG, "onConnectedFailed");
                    synchronized (LOCK) {
                        mCurrentDeviceKey = null;
                        setConnectState(new StateIdle());
                        String errorMessage = "Unknown error";
                        if (null != e) {
                            errorMessage = e.getMessage();
                        }
                        int errorCode = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_UNKNOWN.ordinal();
                        if (KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE == code) {
                            errorCode = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE.ordinal();
                        } else if (KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_TIMEOUT == code) {
                            errorCode = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_TIMEOUT.ordinal();
                        } else if (KageSocket.ISocketCallback.CONNECT_ERROR_CODE_HAND_SHAKE_NOT_COMPLETE == code) {
                            errorCode = ConnectFailedReason.CONNECT_ERROR_CODE_HAND_SHAKE_UNDONE.ordinal();
                        } else if (KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_UNKNOWN == code) {
                            errorCode = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_UNKNOWN.ordinal();
                        }
                        final int finalErrorCode = errorCode;
                        final String finalErrorMessage = errorMessage;
                        mHandler.post(() -> notifyDeviceConnectFailed(device, finalErrorCode, finalErrorMessage));
                    }
                }

                @Override
                public void onConnected() {
                    Log.d(TAG, "onConnectedSuccess");
                    synchronized (LOCK) {
                        PromptPhoneConnectedCommand promptPhoneConnectedCommand = new PromptPhoneConnectedCommand();
                        promptPhoneConnectedCommand.localIp = mConfig.localHost;
                        promptPhoneConnectedCommand.phoneName = mConfig.name;
                        device.sendCommand(promptPhoneConnectedCommand);
                        setConnectState(new DeviceManager.StateConnected(device));
                        notifyDeviceChangedToProxy(device);
                        mHandler.post(() -> notifyDeviceConnected(device));
                    }
                }

                @Override
                public void onDisConnect() {
                    synchronized (LOCK) {
                        mCurrentDeviceKey = null;
                        setConnectState(new StateIdle());
                        notifyDeviceDisconnectToProxy(device);
                        mHandler.post(() -> notifyDeviceDisconnect(device));
                        if (null != delayToConnectDeviceInfoIp && null != mDeviceScanner) {
                            Device dev = mDeviceScanner.queryDevice(delayToConnectDeviceInfoIp);
                            if (dev != null) {
                                connectDevice(dev.getDeviceInfo());
                                delayToConnectDeviceInfoIp = null;
                            }
                        }
                    }
                }
            });
            device.connect(timeout);
            mHandler.post(() -> notifyDeviceConnecting(device));
        }

        @Override
        public void disConnect() {

        }
    }

    private void notifyDeviceChangedToProxy(Device device) {
        IProxy[] tempProxyList = null;
        synchronized (LOCK) {
            int size = mProxyList.size();
            if (0 < size) {
                tempProxyList = new IProxy[size];
                mProxyList.toArray(tempProxyList);
            }
        }
        if (null != tempProxyList) {
            for (IProxy iProxy : tempProxyList) {
                iProxy.onDeviceConnected(device);
            }
        }
    }

    private void notifyDeviceDisconnectToProxy(Device device) {
        IProxy[] tempProxyList = null;
        synchronized (LOCK) {
            int size = mProxyList.size();
            if (0 < size) {
                tempProxyList = new IProxy[size];
                mProxyList.toArray(tempProxyList);
            }
        }
        if (null != tempProxyList) {
            for (IProxy iProxy : tempProxyList) {
                iProxy.onDeviceDisconnected(device);
            }
        }
    }

    private class StateConnecting implements ConnectState {
        private Device mDevice;

        StateConnecting(Device device) {
            mDevice = device;
        }

        @Override
        public void connect(DeviceInfo deviceInfo) {

        }

        @Override
        public void connect(DeviceInfo deviceInfo, int timeout) {

        }

        @Override
        public void disConnect() {
            synchronized (LOCK) {
                mCurrentDeviceKey = null;
            }
        }
    }

    private class StateConnected implements ConnectState {
        private Device mDevice;

        StateConnected(Device device) {
            mDevice = device;
        }

        @Override
        public void connect(DeviceInfo deviceInfo) {
            String ip = deviceInfo.getIp();
            if (!mDevice.getIp().equals(ip)) {
                delayToConnectDeviceInfoIp = ip;
                disConnect();
            }
        }

        @Override
        public void connect(DeviceInfo deviceInfo, int timeout) {

        }

        @Override
        public void disConnect() {
            mDevice.disConnect();
        }
    }

    public void addProxy(IProxy proxy) {
        Device currentDevice = getCurrentDevice();
        if (null != currentDevice && currentDevice.isConnected()) {
            proxy.onDeviceConnected(currentDevice);
        }
        synchronized (LOCK) {
            if (!mProxyList.contains(proxy)) {
                mProxyList.add(proxy);
            }
        }
    }

    public void removeProxy(IProxy proxy) {
        synchronized (LOCK) {
            mProxyList.remove(proxy);
        }
    }

    public void sendCommandToCurrentDevice(Command command) {
        Device device = getCurrentDevice();
        if (device != null) {
            device.sendCommand(command);
        }
    }

    interface ConnectState {
        void connect(DeviceInfo deviceInfo);

        /**
         * @param deviceInfo device info
         * @param timeout    the timeout value in milliseconds
         */
        void connect(DeviceInfo deviceInfo, int timeout);

        void disConnect();
    }
}

File: app/src/main/kotlin/com/absinthe/kage/device/DeviceManager.kt
package com.absinthe.kage.device

import android.bluetooth.BluetoothAdapter
import android.os.Build
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import com.absinthe.kage.connect.IProxy
import com.absinthe.kage.connect.tcp.KageSocket
import com.absinthe.kage.device.Device.IConnectCallback
import com.absinthe.kage.device.DeviceScanner.IScanCallback
import com.absinthe.kage.device.cmd.PromptPhoneConnectedCommand
import com.absinthe.kage.device.model.DeviceConfig
import com.absinthe.kage.device.model.DeviceInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

object DeviceManager : KageObservable(), LifecycleObserver {

    private const val DEFAULT_CONNECT_TIMEOUT = 10 * 1000

    private val LOCK = ByteArray(0)
    private val observers: MutableList<IDeviceObserver> = ArrayList()
    private val mProxyList: MutableList<IProxy> = ArrayList()

    private var mDeviceScanner: DeviceScanner = DeviceScanner()
    private var mConnectState: ConnectState = StateIdle()
    private var mCurrentDeviceKey: String? = null
    private var delayToConnectDeviceInfoIp: String? = null

    var config: DeviceConfig = DeviceConfig()
        private set

    var scanPeriod: Int
        get() = mDeviceScanner.scanPeriod
        set(value) {
            mDeviceScanner.scanPeriod = value
        }

    val deviceInfoList: List<DeviceInfo>
        get() {
            val deviceInfoList: MutableList<DeviceInfo> = ArrayList()
            val devices = mDeviceScanner.devices
            for (device in devices.values) {
                val deviceInfo = device.deviceInfo
                if (device.state != DeviceInfo.STATE_IDLE) {
                    deviceInfoList.add(0, deviceInfo)
                } else {
                    deviceInfoList.add(deviceInfo)
                }
            }
            return deviceInfoList
        }

    val currentDeviceInfo: DeviceInfo?
        get() {
            val device = currentDevice ?: return null
            return device.deviceInfo
        }

    val isConnected: Boolean
        get() {
            val currentDevice = currentDevice
            return currentDevice?.isConnected ?: false
        }

    private val currentDevice: Device?
        get() {
            var device: Device?
            synchronized(LOCK) {
                if (null == mCurrentDeviceKey) {
                    return null
                }
                device = mDeviceScanner.queryDevice(mCurrentDeviceKey)
            }
            return device
        }

    fun init() {
        config.name = BluetoothAdapter.getDefaultAdapter().name ?: Build.MODEL
        mDeviceScanner.setConfig(config)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun release() {
        if (isConnected) {
            disConnectDevice()
        }
        mDeviceScanner.stopScan()
    }

    fun startMonitorDevice() {
        startMonitorDevice(scanPeriod)
    }

    /**
     * 停止监测设备
     */
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun stopMonitorDevice() {
        mDeviceScanner.stopScan()
    }

    /**
     * 连接指定的设备，已连接事件会通知Observer。
     */
    fun connectDevice(deviceInfo: DeviceInfo) {
        synchronized(LOCK) {
            mConnectState.connect(deviceInfo)
        }
    }

    /**
     * @param deviceInfo Device info
     * @param timeout    连接超时时间，单位为毫秒
     */
    @Synchronized
    fun connectDevice(deviceInfo: DeviceInfo, timeout: Int) {
        mConnectState.connect(deviceInfo, timeout)
    }

    /**
     * 如果已连接设备，断开当前连接。已断开事件会通知Observer
     */
    fun disConnectDevice() {
        mConnectState.disConnect()
    }

    fun onlineDevice(deviceInfo: DeviceInfo): DeviceInfo? {
        return mDeviceScanner.onlineDevice(deviceInfo)
    }

    fun offlineDevice(deviceInfo: DeviceInfo?): Boolean {
        return mDeviceScanner.offlineDevice(deviceInfo)
    }

    /**
     * 开始监测设备
     *
     * @param period period
     * @return resultCode
     */
    private fun startMonitorDevice(period: Int): Int {
        val scanResult = mDeviceScanner.startScan(period, object : IScanCallback {
            override fun onDeviceOnline(device: Device?) {
                GlobalScope.launch(Dispatchers.Main) {
                    notifyFindDevice(device)
                }
            }

            override fun onDeviceOffline(device: Device?) {
                GlobalScope.launch(Dispatchers.Main) {
                    notifyLostDevice(device)
                }
            }

            override fun onDeviceInfoChanged(device: Device?) {
                GlobalScope.launch(Dispatchers.Main) {
                    notifyDeviceInfoChanged(device)
                }
            }

            override fun onDeviceNotice(device: Device?) {
                GlobalScope.launch(Dispatchers.Main) {
                    notifyDeviceNotice(device)
                }
            }
        })
        return if (scanResult) {
            Result.RESULT_START_MONITOR_DEVICE_SUCCESS.ordinal
        } else {
            Result.RESULT_START_MONITOR_DEVICE_FAILED_INTERNAL_ERROR.ordinal
        }
    }

    @Synchronized
    override fun register(observer: IDeviceObserver?) {
        if (observer == null) {
            return
        }
        var hasRegistered = false
        val size = observers.size
        for (i in 0 until size) {
            if (observer == observers[i]) {
                hasRegistered = true
                break
            }
        }
        if (!hasRegistered) {
            observers.add(observer)
        }
    }

    @Synchronized
    override fun unregister(observer: IDeviceObserver?) {
        if (observer == null) {
            return
        }
        observers.remove(observer)
    }

    override fun notifyFindDevice(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            observer.onFindDevice(deviceInfo)
        }
    }

    override fun notifyLostDevice(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            observer.onLostDevice(deviceInfo)
        }
    }

    override fun notifyDeviceConnected(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            deviceInfo.state = DeviceInfo.STATE_CONNECTED
            observer.onDeviceConnected(deviceInfo)
        }
    }

    override fun notifyDeviceDisconnect(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            deviceInfo.state = DeviceInfo.STATE_IDLE
            observer.onDeviceDisConnect(deviceInfo)
        }
    }

    override fun notifyDeviceConnectFailed(device: Device?, errorCode: Int, errorMessage: String?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            observer.onDeviceConnectFailed(deviceInfo, errorCode, errorMessage)
        }
    }

    private fun notifyDeviceInfoChanged(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            observer.onDeviceInfoChanged(deviceInfo)
        }
    }

    private fun notifyDeviceConnecting(device: Device) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device.deviceInfo
            observer.onDeviceConnecting(deviceInfo)
        }
    }

    private fun notifyDeviceNotice(device: Device?) {
        var localObservers: Array<IDeviceObserver>
        synchronized(this) { localObservers = observers.toTypedArray() }
        for (observer in localObservers) {
            val deviceInfo = device!!.deviceInfo
            observer.onDeviceNotice(deviceInfo)
        }
    }

    private fun setConnectState(connectState: ConnectState) {
        mConnectState = connectState
    }

    private class StateIdle : ConnectState {
        override fun connect(deviceInfo: DeviceInfo) {
            connect(deviceInfo, DEFAULT_CONNECT_TIMEOUT)
        }

        override fun connect(deviceInfo: DeviceInfo, timeout: Int) {
            var t = timeout
            Timber.d("StateIdle connect")

            if (timeout < 0) {
                t = DEFAULT_CONNECT_TIMEOUT
            }

            val key = deviceInfo.ip
            val device = mDeviceScanner.queryDevice(key) ?: return

            synchronized(LOCK) { mCurrentDeviceKey = key }
            setConnectState(StateConnecting())
            device.setConnectCallback(object : IConnectCallback {

                override fun onConnectedFailed(errorCode: Int, e: Exception?) {
                    Timber.d("onConnectedFailed")
                    synchronized(LOCK) {
                        mCurrentDeviceKey = null
                        setConnectState(StateIdle())
                        var errorMessage: String? = "Unknown error"

                        if (e != null) {
                            errorMessage = e.message
                        }
                        var code = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_UNKNOWN.ordinal

                        when (errorCode) {
                            KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE -> {
                                code = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE.ordinal
                            }
                            KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_TIMEOUT -> {
                                code = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_TIMEOUT.ordinal
                            }
                            KageSocket.ISocketCallback.CONNECT_ERROR_CODE_HAND_SHAKE_NOT_COMPLETE -> {
                                code = ConnectFailedReason.CONNECT_ERROR_CODE_HAND_SHAKE_UNDONE.ordinal
                            }
                            KageSocket.ISocketCallback.CONNECT_ERROR_CODE_CONNECT_UNKNOWN -> {
                                code = ConnectFailedReason.CONNECT_ERROR_CODE_CONNECT_UNKNOWN.ordinal
                            }
                        }

                        GlobalScope.launch(Dispatchers.Main) {
                            notifyDeviceConnectFailed(device, code, errorMessage)
                        }
                    }
                }

                override fun onConnected() {
                    Timber.d("onConnectedSuccess")
                    synchronized(LOCK) {
                        device.sendCommand(PromptPhoneConnectedCommand().apply {
                            localIp = config.localHost
                            phoneName = config.name
                        })
                        setConnectState(StateConnected(device))
                        notifyDeviceChangedToProxy(device)

                        GlobalScope.launch(Dispatchers.Main) {
                            notifyDeviceConnected(device)
                        }
                    }
                }

                override fun onDisConnect() {
                    synchronized(LOCK) {
                        mCurrentDeviceKey = null
                        setConnectState(StateIdle())
                        notifyDeviceDisconnectToProxy(device)
                        GlobalScope.launch(Dispatchers.Main) {
                            notifyDeviceDisconnect(device)
                        }

                        delayToConnectDeviceInfoIp?.let { ip ->
                            mDeviceScanner.queryDevice(ip)?.let {
                                connectDevice(it.deviceInfo)
                                delayToConnectDeviceInfoIp = null
                            }
                        }
                    }
                }
            })
            device.connect(t)
            GlobalScope.launch(Dispatchers.Main) {
                notifyDeviceConnecting(device)
            }
        }

        override fun disConnect() {}
    }

    private fun notifyDeviceChangedToProxy(device: Device) {
        for (iProxy in mProxyList) {
            iProxy.onDeviceConnected(device)
        }
    }

    private fun notifyDeviceDisconnectToProxy(device: Device) {
        for (iProxy in mProxyList) {
            iProxy.onDeviceDisconnected(device)
        }
    }

    private class StateConnecting internal constructor() : ConnectState {

        override fun connect(deviceInfo: DeviceInfo) {}

        override fun connect(deviceInfo: DeviceInfo, timeout: Int) {}

        override fun disConnect() {
            synchronized(LOCK) { mCurrentDeviceKey = null }
        }
    }

    private class StateConnected internal constructor(private val mDevice: Device) : ConnectState {

        override fun connect(deviceInfo: DeviceInfo) {
            val ip = deviceInfo.ip
            if (mDevice.ip != ip) {
                delayToConnectDeviceInfoIp = ip
                disConnect()
            }
        }

        override fun connect(deviceInfo: DeviceInfo, timeout: Int) {}

        override fun disConnect() {
            mDevice.disConnect()
        }

    }

    fun addProxy(proxy: IProxy) {
        val currentDevice = currentDevice
        if (null != currentDevice && currentDevice.isConnected) {
            proxy.onDeviceConnected(currentDevice)
        }
        synchronized(LOCK) {
            if (!mProxyList.contains(proxy)) {
                mProxyList.add(proxy)
            }
        }
    }

    fun removeProxy(proxy: IProxy) {
        synchronized(LOCK) { mProxyList.remove(proxy) }
    }

    fun sendCommandToCurrentDevice(command: Command?) {
        val device = currentDevice
        device?.sendCommand(command!!)
    }

    enum class ConnectFailedReason {
        CONNECT_ERROR_CODE_CONNECT_UNKNOWN,  //未知原因
        CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE,  //远端 IP 或 Port 不可到达
        CONNECT_ERROR_CODE_CONNECT_TIMEOUT,  //连接超时
        CONNECT_ERROR_CODE_HAND_SHAKE_UNDONE //协议握手未完成
    }

    enum class Result {
        RESULT_START_MONITOR_DEVICE_SUCCESS,
        RESULT_START_MONITOR_DEVICE_FAILED_NOT_INIT,
        RESULT_START_MONITOR_DEVICE_FAILED_UNAUTHORIZED,
        RESULT_START_MONITOR_DEVICE_FAILED_INTERNAL_ERROR
    }

    internal interface ConnectState {
        fun connect(deviceInfo: DeviceInfo)

        /**
         * @param deviceInfo device info
         * @param timeout    the timeout value in milliseconds
         */
        fun connect(deviceInfo: DeviceInfo, timeout: Int)
        fun disConnect()
    }
}